pkg/helm/urlfetcher.go

```diff
@@ -0,0 +1,155 @@
+package helm
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"istio.io/istio/pkg/log"
+	"k8s.io/helm/pkg/chartutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type FileDownloader struct {
+	// HTTP/HTTPS Client
+	client *http.Client
+}
+
+type URLFetcher struct {
+	// url to download the charts
+	url string
+	// url to download the verification file
+	verifyUrl string
+	// whether to verify downloaded tar
+	verify bool
+	// whether need to untar
+	untar bool
+	// dir of charts downloaded to, empty as default to temp dir
+	destDir string
+	// untar destination
+	untarDir string
+	// downloader
+	downloader *FileDownloader
+}
+
+// fetch charts and verify with sha file
+func (f *URLFetcher) fetchChartAndVerify(shaF string) error {
+	c := f.downloader
+
+	// fetch to default temp dir
+	saved, err := c.DownloadTo(f.url, f.destDir)
+	if err != nil {
+		return err
+	}
+	file, err := os.Open(saved)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	if f.verify {
+		// verify with sha file
+		hashAll, err := ioutil.ReadFile(shaF)
+		if err != nil {
+			return fmt.Errorf("failed to read sha file: %s", err)
+		}
+		hash := strings.Fields(string(hashAll))[0]
+		// Verify downloaded tar
+		h := sha256.New()
+		if _, err := io.Copy(h, file); err != nil {
+			log.Error(err.Error())
+		}
+		sum := h.Sum(nil)
+		if !strings.EqualFold(hex.EncodeToString(sum), hash) {
+			return errors.New("checksum does not match")
+		}
+	}
+
+	// After verification, untar the chart into the requested directory.
+	if f.untar {
+		ud := f.untarDir
+		if !filepath.IsAbs(ud) {
+			ud = filepath.Join(f.destDir, ud)
+		}
+		if fi, err := os.Stat(ud); err != nil {
+			if err := os.MkdirAll(ud, 0755); err != nil {
+				return fmt.Errorf("failed to untar (mkdir): %s", err)
+			}
+
+		} else if !fi.IsDir() {
+			return fmt.Errorf("failed to untar: %s is not a directory", ud)
+		}
+
+		return chartutil.ExpandFile(ud, saved)
```

Comment:
I would just do a normal untar operation then. Actually I think it may depend on what format we expect when we render manifest based on local charts.